source/Calamari/Kubernetes/ResourceStatus/ResourceRetriever.cs

```diff
@@ -0,0 +1,73 @@
+using System.Collections.Generic;
+using System.Linq;
+using Calamari.Kubernetes.Integration;
+using Calamari.Kubernetes.ResourceStatus.Resources;
+
+namespace Calamari.Kubernetes.ResourceStatus
+{
+    /// <summary>
+    /// Retrieves resources information from a kubernetes cluster
+    /// </summary>
+    public interface IResourceRetriever
+    {
+        /// <summary>
+        /// Gets the resources identified by the resourceIdentifiers and all their owned resources
+        /// </summary>
+        IEnumerable<Resource> GetAllOwnedResources(IEnumerable<ResourceIdentifier> resourceIdentifiers, Kubectl kubectl);
+    }
+    
+    public class ResourceRetriever : IResourceRetriever
+    {
+        private readonly IKubectlGet kubectlGet;
+
+        public ResourceRetriever(IKubectlGet kubectlGet)
+        {
+            this.kubectlGet = kubectlGet;
+        }
+        
+        /// <inheritdoc />
+        public IEnumerable<Resource> GetAllOwnedResources(IEnumerable<ResourceIdentifier> resourceIdentifiers, Kubectl kubectl)
+        {
+            var resources = resourceIdentifiers
+                .Select(identifier => GetResource(identifier, kubectl))
+                .Where(resource => resource != null)
+                .ToList();
+
+            foreach (var resource in resources)
+            {
+                EnrichWithChildren(resource, kubectl);
+            }
+
+            return resources;
+        }
+
+        private void EnrichWithChildren(Resource resource, Kubectl kubectl)
+        {
+            var children = GetChildrenResources(resource, kubectl).ToList();
+            foreach (var child in children)
+            {
+                EnrichWithChildren(child, kubectl);
+            }
+            resource.Children = children;
+        }
+    
+        private Resource GetResource(ResourceIdentifier resourceIdentifier, Kubectl kubectl)
+        {
+            var result = kubectlGet.Resource(resourceIdentifier.Kind, resourceIdentifier.Name, resourceIdentifier.Namespace, kubectl);
+            return string.IsNullOrEmpty(result) ? null : ResourceFactory.FromJson(result);
+        }
+    
+        private IEnumerable<Resource> GetChildrenResources(Resource parentResource, Kubectl kubectl)
+        {
+            var childKind = parentResource.ChildKind;
+            if (string.IsNullOrEmpty(childKind))
+            {
+                return Enumerable.Empty<Resource>();
+            }
+
+            var result = kubectlGet.AllResources(childKind, parentResource.Namespace, kubectl);
+            var resources = ResourceFactory.FromListJson(result);
+            return resources.Where(resource => resource.OwnerUids.Contains(parentResource.Uid)).ToList();
+        }
```

Comment:
Moving the assignment into where the resource is created will allow us to cut out the `EnrichWithChildren` method
```suggestion
            return resources.Where(resource => resource.OwnerUids.Contains(parentResource.Uid))
                            .Select(child =>
                                    {
                                        child.Children = GetChildrenResources(child, kubectl);
                                        return child;
                                    })
                            .ToList();
```

line 30 above then also just becomes
```
            foreach (var resource in resources)
            {
                resource.Children = GetChildrenResources(resource, kubectl);
            }
```

Generally I like to avoid objects being modified outside the scope of their creation when possible so we can consider them immutable outside that context.